File: src/main/java/tomihi/atomiki/models/CompressedUserGame.java
package tomihi.atomiki.models;

import lombok.Data;
import org.springframework.lang.NonNull;
import tomihi.atomiki.game.*;

@Data
public class CompressedUserGame {
    private Status status;
    @NonNull
    private CompressedMovesLog movesLog;
    @NonNull
    private CompressedField field;
    private CompetitorMarks competitorMarks;

    private CompressedUserGame(UserGame userGame) {
        this.status = userGame.getStatus();
        this.movesLog = CompressedMovesLog.fromMovesLog(userGame.getMovesLog());
        this.field = CompressedField.fromField(userGame.getField());
        this.competitorMarks = userGame.getCompetitorMarks();
    }

    public CompressedUserGame() {
    }

    public static CompressedUserGame fromUserGame(UserGame userGame) {
        return new CompressedUserGame(userGame);
    }

    public UserGame toUserGame(GameSettings gameSettings) throws WrongActionForCoords, ImpossibleAtomLocationException, AtomsOverflowException {
        return new UserGame(this.status, this.movesLog.toMovesLog(), this.field.toField(gameSettings), this.competitorMarks);
    }
}

File: app/src/types/transport/CompressedUserGame.ts
import {Status} from "./Status";
import {CompressedMovesLog} from "./CompressedMovesLog";
import {CompressedField} from "./CompressedField";
import {CompetitorMarks} from "./CompetitorMarks";

export interface CompressedUserGame {
    status: Status;
    movesLog: CompressedMovesLog;
    field: CompressedField;
    competitorMarks?: CompetitorMarks;
}